src/bc.hpp

```diff
@@ -0,0 +1,123 @@
+                                ? bounds.ks(parthenon::IndexDomain::interior)
+                                : bounds.ks(parthenon::IndexDomain::entire),
+                            IsDomainBound(pmb, parthenon::BoundaryFace::outer_x3)
+                                ? bounds.ke(parthenon::IndexDomain::interior)
+                                : bounds.ke(parthenon::IndexDomain::entire)}};
+}
+
+enum class BCSide { Inner, Outer };
+enum class BCType { Outflow, Reflect };
+
+template <parthenon::CoordinateDirection DIR, BCSide SIDE, BCType TYPE>
+void ApplyBC(parthenon::MeshBlock *pmb, parthenon::VariablePack<Real> &q,
+             parthenon::IndexRange &nvar, const bool is_normal, const bool coarse) {
+  // convenient shorthands
+  constexpr bool X1 = (DIR == parthenon::X1DIR);
+  constexpr bool X2 = (DIR == parthenon::X2DIR);
+  constexpr bool X3 = (DIR == parthenon::X3DIR);
+  constexpr bool INNER = (SIDE == BCSide::Inner);
+
+  constexpr parthenon::BoundaryFace bface =
+      INNER ? (X1 ? parthenon::BoundaryFace::inner_x1
+                  : (X2 ? parthenon::BoundaryFace::inner_x2
+                        : parthenon::BoundaryFace::inner_x3))
+            : (X1 ? parthenon::BoundaryFace::outer_x1
+                  : (X2 ? parthenon::BoundaryFace::outer_x2
+                        : parthenon::BoundaryFace::outer_x3));
+
+  // check that we are actually on a physical boundary
+  if (!IsDomainBound(pmb, bface)) {
+    return;
+  }
+
+  const auto &bounds = coarse ? pmb->c_cellbounds : pmb->cellbounds;
+
+  const auto &range = X1 ? bounds.GetBoundsI(parthenon::IndexDomain::interior)
+                         : (X2 ? bounds.GetBoundsJ(parthenon::IndexDomain::interior)
+                               : bounds.GetBoundsK(parthenon::IndexDomain::interior));
+  const int ref = INNER ? range.s : range.e;
+
+  std::string label = (TYPE == BCType::Reflect ? "Reflect" : "Outflow");
+  label += (INNER ? "Inner" : "Outer");
+  label += "X" + std::to_string(DIR);
+
+  constexpr parthenon::IndexDomain domain =
+      INNER ? (X1 ? parthenon::IndexDomain::inner_x1
+                  : (X2 ? parthenon::IndexDomain::inner_x2
+                        : parthenon::IndexDomain::inner_x3))
+            : (X1 ? parthenon::IndexDomain::outer_x1
+                  : (X2 ? parthenon::IndexDomain::outer_x2
+                        : parthenon::IndexDomain::outer_x3));
+
+  // used for reflections
+  const int offset = 2 * ref + (INNER ? -1 : 1);
+
+  pmb->par_for_bndry(
+      label, nvar, domain, coarse,
+      KOKKOS_LAMBDA(const int &l, const int &k, const int &j, const int &i) {
+        if (!q.IsAllocated(l)) return;
+        if (TYPE == BCType::Reflect) {
+          q(l, k, j, i) =
+              (is_normal ? -1.0 : 1.0) *
+              q(l, X3 ? offset - k : k, X2 ? offset - j : j, X1 ? offset - i : i);
+        } else {
+          q(l, k, j, i) = q(l, X3 ? ref : k, X2 ? ref : j, X1 ? ref : i);
```

Comment:
I might wonder about executing these ternary statements inside the kernel and wonder if the CUDA compiler is smart enough to compile them out.

However, this kernel is certainly memory bandwidth limited so the extra integer computes are irrelevant